packages/docusaurus-theme-common/src/utils/codeBlockUtils.ts

```diff
@@ -147,32 +175,120 @@ function getAllMagicCommentDirectiveStyles(
   }
 }
 
-export function parseCodeBlockTitle(metastring?: string): string {
-  return metastring?.match(codeBlockTitleRegex)?.groups!.title ?? '';
+/**
+ * Rewrites the range syntax to special options. e.g.
+ * `{1,2,3-4,5} => highlight=1 highlight=2 highlight=3-4 highlight=5`
+ * @param metastring The input metastring with the range syntax
+ * @returns The string where the range syntax has been rewritten
+ */
+function rewriteLinesRange(metastring: string): string {
+  metastringLinesRangeRegex.lastIndex = 0;
+
+  return metastring.replaceAll(metastringLinesRangeRegex, (_, range) => {
+    return (range as string)
+      .split(',')
+      .map((r) => `${highlightOptionKey}=${r}`)
+      .join(' ');
+  });
 }
 
-function getMetaLineNumbersStart(metastring?: string): number | undefined {
-  const showLineNumbersMeta = metastring
-    ?.split(' ')
-    .find((str) => str.startsWith('showLineNumbers'));
-
-  if (showLineNumbersMeta) {
-    if (showLineNumbersMeta.startsWith('showLineNumbers=')) {
-      const value = showLineNumbersMeta.replace('showLineNumbers=', '');
-      return parseInt(value, 10);
+function parseCodeBlockOptions(
```

Comment:
Will need to review more in depth but that looks quite complicated

I'd prefer if we start by centralizing parsing the options, but keep the old methods like `getMetaLineNumbersStart`

We can eventually rewrite the parser later in another PR, but I'd prefer is we don't refactor everything at once.

Please split this into multiple sequential PRs:
- first centralize parsing
- then refactor how we parse

Not everything at once

Reply:
IMO Its fairly straight forward considering what docusaurus needs to support. If we would start from scratch, we could add some more restrictions. But this code ensures backwards compatibility with a variety of option syntaxes (no options, plain values etc. 

1. `Highlight` needs special treatment for backwards compatibilty of `lineClassNames`
2. Options without values are used for markers like `showLineNumbers`
3. Boolean values are new, but influenced by the linked PRs in [#11008](https://github.com/facebook/docusaurus/issues/11008#issuecomment-2736770049) to be able to enable/disable some options. 
4. Number values are for `showLineNumbers` compatibility. 

I'll try to split apart the value handling a bit to reduce the complexity of this single function.

Reply:
Before we even think about changing the way we parse things, I'd prefer to keep the exact same parsing logic as before, extract it to a single place, and cover it widely with unit tests.

Once this is well-covered, you can start changing the implementation, ensuring all the tests keep passing.

Reply:
#11017 and #11019 refactor and cleanup the parsing logic and add more tests to the current expectations. Then only afterwards in #11021 the parsing is adjusted.